File: src/compiler/typescript_to_rust/generators/ic_object/functions/stable_read.ts
import { Rust } from '../../../../../types';

export function generateIcObjectFunctionStableRead(): Rust {
    // TODO Optimize vec conversion https://github.com/demergent-labs/azle/issues/482
    return /* rust */ `
        fn _azle_ic_stable_read(
            _this: &boa_engine::JsValue,
            _aargs: &[boa_engine::JsValue],
            _context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let offset: u32 = _aargs
                .get(0)
                .unwrap()
                .clone()
                .azle_try_from_js_value(_context)
                .unwrap();
            let length: u32 = _aargs
                .get(1)
                .unwrap()
                .clone()
                .azle_try_from_js_value(_context)
                .unwrap();

            let mut buf: Vec<u8> = vec![0; length as usize];
            ic_cdk::api::stable::stable_read(offset, &mut buf);
            Ok(buf.azle_into_js_value(_context))
        }
    `;
}

File: src/compiler/typescript_to_rust/azle_generate/src/body/ic_object/functions/stable_read.rs
pub fn generate() -> proc_macro2::TokenStream {
    quote::quote! {
        fn stable_read(
            _this: &boa_engine::JsValue,
            aargs: &[boa_engine::JsValue],
            context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let offset: u32 = aargs
                .get(0)
                .ok_or_else(|| "An argument for 'offset' was not provided".to_js_error(None))?
                .clone()
                .try_from_vm_value(&mut *context)?;

            let length: u32 = aargs
                .get(1)
                .ok_or_else(|| "An argument for 'length' was not provided".to_js_error(None))?
                .clone()
                .try_from_vm_value(&mut *context)?;

            let mut buf: Vec<u8> = vec![0; length as usize];
            ic_cdk::api::stable::stable_read(offset, &mut buf);

            buf.to_vec()
                .try_into_vm_value(context)
                .map_err(|vmc_err| vmc_err.to_js_error(None))
        }
    }
}
